From 2029ec72d680fc77349c241ed9ddb08455ae97e6 Mon Sep 17 00:00:00 2001
From: Vaibhav Hiremath <hvaibhav@ti.com>
Date: Sun, 27 Nov 2011 00:00:41 +0530
Subject: [PATCH 0421/1092] da8xx: Fix GLCD initialization sequence

Sometimes system comes up with a 'white screen' on the LCD after
boot, and nothing is ever displayed. This was consistent with the
'reset' button (or linux reboot). This is fixed by switching off
and switching on the panel power and backlight after enabling
raster. Earlier this sequence was done before enabling the raster.

Signed-off-by: Rajashekhara, Sudhakar <sudhakar.raj@ti.com>
Signed-off-by: Vaibhav Hiremath <hvaibhav@ti.com>
---
 arch/arm/mach-davinci/board-da850-evm.c | 15 +++++----------
 drivers/video/da8xx-fb.c                |  8 ++++++++
 2 files changed, 13 insertions(+), 10 deletions(-)

diff --git a/arch/arm/mach-davinci/board-da850-evm.c b/arch/arm/mach-davinci/board-da850-evm.c
index 6659a90..c5d808a 100644
--- a/arch/arm/mach-davinci/board-da850-evm.c
+++ b/arch/arm/mach-davinci/board-da850-evm.c
@@ -32,7 +32,6 @@
 #include <linux/spi/spi.h>
 #include <linux/spi/flash.h>
 #include <linux/delay.h>
-#include <linux/wl12xx.h>
 
 #include <asm/mach-types.h>
 #include <asm/mach/arch.h>
@@ -794,11 +793,13 @@ static const short da850_evm_mmcsd0_pins[] __initconst = {
 
 static void da850_panel_power_ctrl(int val)
 {
-	/* lcd backlight */
-	gpio_set_value(DA850_LCD_BL_PIN, val);
-
 	/* lcd power */
 	gpio_set_value(DA850_LCD_PWR_PIN, val);
+
+	mdelay(200);
+
+	/* lcd backlight */
+	gpio_set_value(DA850_LCD_BL_PIN, val);
 }
 
 static int da850_lcd_hw_init(void)
@@ -818,12 +819,6 @@ static int da850_lcd_hw_init(void)
 	gpio_direction_output(DA850_LCD_BL_PIN, 0);
 	gpio_direction_output(DA850_LCD_PWR_PIN, 0);
 
-	/* Switch off panel power and backlight */
-	da850_panel_power_ctrl(0);
-
-	/* Switch on panel power and backlight */
-	da850_panel_power_ctrl(1);
-
 	return 0;
 }
 
diff --git a/drivers/video/da8xx-fb.c b/drivers/video/da8xx-fb.c
index f550811..a834613 100644
--- a/drivers/video/da8xx-fb.c
+++ b/drivers/video/da8xx-fb.c
@@ -994,6 +994,14 @@ static int fb_wait_for_vsync(struct fb_info *info)
 	if (ret == 0)
 		return -ETIMEDOUT;
 
+	if (par->panel_power_ctrl) {
+		/* Switch off panel power and backlight */
+		par->panel_power_ctrl(0);
+
+		/* Switch on panel power and backlight */
+		par->panel_power_ctrl(1);
+	}
+
 	return 0;
 }
 
-- 
1.7.11.2

